### ISSUE TITLE
MMU streaming on top of catalogue stream

### ISSUE THREAD

**Issue by Smith42** [astronomy-commons/lsdb#1576]
# Copied excerpt from the LSBD Training Data Delivery for AI Models document below:

TL;DR we want to design a streaming protocol that allows HATS dataset crossmatching with the end goal of being able to pre-train a ML model at a sufficient MFU rate.

## Multimodal foundation-model pretraining.

Train an astro-foundation model on aligned (images, spectrum, light curves) tuples, spanning an anchor modality (e.g. huge image catalog) and multiple extra catalogs with other modalities (images, spectra, text, tabular). In order to maximize throughput, the cross-matching algorithm may keep just a few extra catalogs per anchor row. That would require a metric to maximize the information per row. There may be other hacks to maximize throughput here. We want to leverage the already-existing catalog infinite stream function.

Requirements:
- Deliver batches of cross-modal samples
- Sustain GPU-bound throughput on a single node (Perlumutter or alternatives)
- Contiguous training over shuffled data. (Quasi-shuffling by selection Healpix tiles is ok.)
- Build a map catalog of weights (e.g. of match densities, based on pre-existing coverage maps) per crossmatched partition that can be used to inform healpix selection.
- For each Healpix tile make a decision which extra catalogs to keep based on a skymap of pairwise (anchor + extra) weight. This decision may be probabilistic and should be re-evaluated when streaming this partition next time.

Good to have:
- Resume from a checkpoint without repeating data
- Stream tokenized data directly
